### PR TITLE
Update the plotting code

### PR DESCRIPTION
**Pull request by ahartikainen**
#### Summary:

Update the plotting code. 

New faster kde backend.
Added new plots: forest and parcoords 

#### Intended Effect:

Call plotting code either from `fit` or `pystan.plots`

#### Side Effects:

Remove old pymc3 code

#### Documentation:

Needs updating.

#### Reviewer Suggestions: 

@ariddell 

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Ari Hartikainen

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: GPLv3 (http://opensource.org/licenses/GPL-3.0)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by ahartikainen** [comment]
Needs still some updating.

File: pystan/plots/__init__.py [hunk 1]
```diff
@@ -0,0 +1,2 @@
+from .plots import *
+from .plots_data import *
```
Comment (ahartikainen) [inline pystan/plots/__init__.py]: this is a bad habit, import specific functions

File: pystan/plots/plots.py [hunk 2]
```diff
@@ -0,0 +1,621 @@
...
+    """
+    try:
+        import matplotlib.pyplot as plt
+    except ImportError:
+        logger.critical("matplotlib required for plotting.")
+        raise
+    if pars is None:
+        pars = fit.model_pars
+    elif isinstance(pars, string_types):
+        pars = [pars]
```
Comment (ahartikainen) [inline pystan/plots/plots.py]: Add test to check that pars are real pars + enable diagnostic "parameters"

File: pystan/plots/plots.py [hunk 3]
```diff
@@ -0,0 +1,621 @@
...
+    if density:
+        trace_col = 1
+        m = 2
+    else:
+        trace_col = 0
+        m = 1
+    # raise an exception if too many subplots
+    force = kwargs.pop('force', False)
+    if not force:
+        if n > 15:
```
Comment (ahartikainen) [inline pystan/plots/plots.py]: Should we do this?

File: pystan/plots/plots.py [hunk 4]
```diff
@@ -0,0 +1,621 @@
...
+            err_msg = "Too many subplots: (row={},col={})->total={},"\
+                      " please use 'force' keyword to force"\
+                      " enable plotting".format(n, m, n*m)
+            logger.critical(err_msg)
+            raise ValueError(err_msg)
+    # create figure object
+    figsize = kwargs.pop('figsize', (6, max(4.5, 2.5*n)))
```
Comment (ahartikainen) [inline pystan/plots/plots.py]: better defaults

File: pystan/plots/plots.py [hunk 5]
```diff
@@ -0,0 +1,621 @@
...
+    tight_layout = kwargs.pop('tight_layout', True)
+
+    statistic = kwargs.pop('statistic', False)
+    if isinstance(statistic, int) and statistic:
+        statistic = [{'func' : np.mean}]
+    elif isinstance(statistic, str) and statistic:
+        if statistic == 'mean':
+            statistic = [{'func' : np.mean}]
+        elif statistic == 'median':
+            statistic = [{'func' : np.median}]
```
Comment (ahartikainen) [inline pystan/plots/plots.py]: Enable dictionary for one `statistic` --> check that kw func is in the dict.

File: pystan/plots/plots.py [hunk 6]
```diff
@@ -0,0 +1,621 @@
...
+                                                traceplot_dict['vec'],
+                                                axes[ax_row, density_col],\
+                                                plot_dict=statistic_dict, zero_level=0,\
+                                                method='hist')
+        # plot trace
+        vec = traceplot_dict['vec']
+        axes[ax_row,trace_col].plot(vec)
+    # update axis labels
+    axes[ax_row, density_col].set_title(parameter)
+    axes[ax_row, trace_col].set_title(parameter)
```
Comment (ahartikainen) [inline pystan/plots/plots.py]: Make these titles unique

File: pystan/plots/plots.py [hunk 7]
```diff
@@ -0,0 +1,621 @@
...
+    axes[ax_row, density_col].set_ylabel('density')
+    axes[ax_row, trace_col].set_xlabel('trace')
```
Comment (ahartikainen) [inline pystan/plots/plots.py]: What are the correct xlabels/ylabels. Do we need other than these.

File: pystan/plots/plots.py [hunk 8]
```diff
@@ -0,0 +1,621 @@
...
+    if legend:
+        axes[ax_row, density_col].legend()
```
Comment (ahartikainen) [inline pystan/plots/plots.py]: Make sure that legend is meaningful.

File: pystan/plots/plots_data.py [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+    data : dict
+        dict contains
+            par, parameter name
+            name, parameter component name
+            vec,
+            hist, histogram data, optional
+            kde, kde data, optional
+    """
+    # TODO: Add option to plot chains independently
+    # This can be done with np.ndindex tricks
```
Comment (ahartikainen) [inline pystan/plots/plots_data.py]: `np.ndindex` only works with `C-order`

File: pystan/plots/plots_data.py [hunk 10]
```diff
@@ -0,0 +1,332 @@
...
+                else:
+                    x_kde, y_kde = kde_data(vec, limits, c=c)
+                    yield {'par': par,
+                           'name' : name,
+                           'vec' : vec,
+                           'kde' : (x_kde, y_kde)}
+            else:
+                yield {'par' : par,
+                       'name' : name,
+                       'vec' : vec}
```
Comment (ahartikainen) [inline pystan/plots/plots_data.py]: yield or return data in one step?

File: pystan/plots/plots_data.py [hunk 11]
```diff
@@ -0,0 +1,332 @@
...
+        vecs = vecs.reshape([vecs.shape[0]*vecs.shape[1]]+list(par_dims), order='F')
+        vecs = vecs.reshape(-1, m, order='F')
+        if transform is not None:
+            if isinstance(transform, str):
+                if transform.lower() in ('min-max', 'minmax'):
+                    vmin, vmax = vecs.min(0), vecs.max(0)
+                    vecs = (vecs - vmin) / (vmax-vmin)
+                elif transform.lower() in ('standard'):
+                    vmean, vstd = vecs.mean(0), vecs.std(0, ddof=1)
+                    vecs = (vecs - vmean) / vstd
```
Comment (ahartikainen) [inline pystan/plots/plots_data.py]: Add inverse cdf normalization. 

I have an approximation for it working, but I want it to work correctly. Currently the problem is that without some clever summing the cumsum will move unique numbers to different values

    # Current situation
    1, 2, 2, 3 --> 1, 3, 5, 8

    # Option 1 (max value)
    1, 2, 2, 3 --> 1, 5, 5, 8

    # Option 2 (mean value)
    1, 2, 2, 3 --> 1, 4, 4, 8

    # Option 3 (min value)
    1, 2, 2, 3 --> 1, 3, 3, 8

I prefer the mean one, notice however that the cumsum is going over the block normally.

File: pystan/plots/plots_data.py [hunk 12]
```diff
@@ -0,0 +1,332 @@
...
+        indices = np.c_[np.unravel_index(np.arange(m), par_dims, order='F')]
```
Comment (ahartikainen) [inline pystan/plots/plots_data.py]: Is there any ready function for this.

File: pystan/plots/plots.py [hunk 13]
```diff
@@ -0,0 +1,621 @@
...
+    ax.set_yticks(height_list)
+    ax.set_yticklabels(name_list)
+
+    if legend:
+        ax.legend()
+    if tight_layout:
+        fig.tight_layout(pad=0.5)
+    return fig, ax
+
+def parcoords(fit, pars=None, transform=None, divergence=None, **kwargs):
```
Comment (ahartikainen) [inline pystan/plots/plots.py]: better name for this function: `parcoordsplot`

**Comment by ahartikainen** [comment]
Here are some example figures

    import pystan
    import matplotlib.pyplot as plt
    plt.style.use('ggplot')
    # eight schools example
    ...
    fit = stan_model.sampling(data = schools_dat)

    fit.plot(type='trace')
    plt.tight_layout(pad=0.5)
    plt.savefig("traceplot_example1", bbox_inches='tight', dpi=300)

traceplot_example1

    fit.plot(type='trace', statistic=True, figsize=(10,10))
    plt.tight_layout(pad=0.5)
    plt.savefig("traceplot_example2", bbox_inches='tight', dpi=300)

traceplot_example2

    fit.plot(type='trace', statistic=True, fill=False)
    plt.tight_layout(pad=0.5)
    plt.savefig("traceplot_example3", bbox_inches='tight', dpi=300)

traceplot_example3

    fit.plot(type='forest')
    plt.tight_layout(pad=0.5)
    plt.savefig("forestplot_example1", bbox_inches='tight', dpi=300)

forestplot_example1

    fit.plot(type='forest', statistic=True)
    plt.tight_layout(pad=0.5)
    plt.xlim(-10,20)
    plt.savefig("forestplot_example2", bbox_inches='tight', dpi=300)

forestplot_example2

    fit.plot(type='forest', fill={'alpha' : 1}, statistic=True, overlap=1.7)
    plt.tight_layout(pad=0.5)
    plt.xlim(-10,20)
    plt.savefig("forestplot_example3", bbox_inches='tight', dpi=300)

forestplot_example3

    fit.plot(type='parcoords', transform='min-max', divergence=True, alpha=0.05, alpha_div=0.3)
    plt.tight_layout(pad=0.5)
    plt.savefig("parcoords_example1", bbox_inches='tight', dpi=300)

parcoords_example1

    fig, ax = fit.plot(type='parcoords', transform='standard', cmap='inferno', alpha=0.1)
    plt.savefig("parcoords_example2", bbox_inches='tight', dpi=300)

parcoords_example2

**Comment by ahartikainen** [comment]
@ariddell do you want this to next release?

I would want make some poll about the properties and figsizes (or just use some defaults for now and change them later).

**Comment by ariddell** [comment]
I think plotting support can go in 2.18 if it's ready. This all looks
great. If you want to poll users that would be good too. Might also be
easy to just copy the pandas plotting defaults.

What about copying the pandas plotting api and method signature(s)? I
think pandas does ``df.plot(kind='plot-type')`` and returns an axis. I
think we should either do that OR copy RStan as closely as possible.

**Comment by hanyas** [comment]
I am having a weird plotting issue depending on the number of data points I feed to my model and number of sampling iteration:

```Traceback (most recent call last):
  File "<input>", line 30, in <module>
  File "stanfit4anon_model_d6904cb5f5a4c1e3c41d23b6fbbb109e_710359492018507600.pyx", line 512, in stanfit4anon_model_d6904cb5f5a4c1e3c41d23b6fbbb109e_710359492018507600.StanFit4Model.plot
  File "/home/hany/miniconda3/lib/python3.6/site-packages/pystan/plots/plotting.py", line 326, in traceplot
    plot_kde(*traceplot_dict['kde'], fill=fill, ax=axes[ax_row, density_col], **kde_kwargs)
  File "/home/hany/miniconda3/lib/python3.6/site-packages/pystan/plots/plotting.py", line 128, in plot_kde
    kde_plot, = ax.plot(x, y, **kwargs)
  File "/home/hany/miniconda3/lib/python3.6/site-packages/matplotlib/__init__.py", line 1717, in inner
    return func(ax, *args, **kwargs)
  File "/home/hany/miniconda3/lib/python3.6/site-packages/matplotlib/axes/_axes.py", line 1372, in plot
    for line in self._get_lines(*args, **kwargs):
  File "/home/hany/miniconda3/lib/python3.6/site-packages/matplotlib/axes/_base.py", line 404, in _grab_next_args
    for seg in self._plot_args(this, kwargs):
  File "/home/hany/miniconda3/lib/python3.6/site-packages/matplotlib/axes/_base.py", line 384, in _plot_args
    x, y = self._xy_from_xy(x, y)
  File "/home/hany/miniconda3/lib/python3.6/site-packages/matplotlib/axes/_base.py", line 243, in _xy_from_xy
    "have shapes {} and {}".format(x.shape, y.shape))
ValueError: x and y must have same first dimension, but have shapes (200,) and (1,)
```
I am on Linux, Python 3.6 and matplotlib 2.1.2. Could this be a matplotlib issue?

**Comment by ahartikainen** [comment]
Thanks for the info. Can you make minimal working example showing the issue? 

It is probably bug in the plotting code. Let's fix it before merging this PR.

**Comment by hanyas** [comment]
Disclaimer: I am new to Stan and PyStan so the error might even be my fault. 

This is an example of trying to infer the mixing and means of a Gaussian mixture given the standard derivation:

```
import pystan
import numpy as np
import matplotlib.pyplot as plt

plt.style.use('seaborn-darkgrid')

gmm_code = """
data {
	int<lower=0> N;
	int<lower=1> K;
	
	vector[N] x;
	
	vector[K] prior_mean;
	cov_matrix[K] prior_cov;
	real sigma;
}

parameters {
	vector[K] mu;	
	real<lower=0, upper=1> alpha;
}

model {
	alpha ~ uniform(0, 1);
	mu ~ multi_normal(prior_mean, prior_cov);
	
	for (n in 1:N) {
		real lp1; real lp2;
		
		lp1 = bernoulli_lpmf(0| alpha) + normal_lpdf(x[n]| mu[1], sigma);
		lp2 = bernoulli_lpmf(1| alpha) + normal_lpdf(x[n]| mu[2], sigma);
		
		target += log_sum_exp(lp1, lp2);
	}
}
"""

sm = pystan.StanModel(model_code=gmm_code)

N = 500

x = np.zeros(N)

for n in range(N):
	z = np.random.binomial(1, 0.7, size=1)
	if z == 0:
		x[n] = np.random.normal(loc=2.0, scale=0.1, size=1)
	else:
		x[n] = np.random.normal(loc=-2.0, scale=0.1, size=1)

gmm_data = {'N': N,
            'K': 2,
            'x': x,
            'prior_mean': np.zeros(2),
            'prior_cov': np.eye(2),
            'sigma': 0.1}

fit = sm.sampling(data=gmm_data, iter=1000, chains=1)

print(fit)

alpha = fit.extract(permuted=True)['alpha']
mu = fit.extract(permuted=True)['mu']

fit.plot()
plt.show()
```

I sometimes also get the following error:

```
Traceback (most recent call last):
  File "<input>", line 26, in <module>
  File "stanfit4anon_model_285c8791d7ccb0277b2f8bb2375bc202_2904228436846562841.pyx", line 512, in stanfit4anon_model_285c8791d7ccb0277b2f8bb2375bc202_2904228436846562841.StanFit4Model.plot
  File "/home/hany/miniconda3/lib/python3.6/site-packages/pystan/plots/plotting.py", line 306, in traceplot
    for i, traceplot_dict in enumerate(traceplot_data(fit, pars, dtypes, **kwargs)):
  File "/home/hany/miniconda3/lib/python3.6/site-packages/pystan/plots/plots_data.py", line 172, in traceplot_data
    indices = np.c_[np.unravel_index(np.arange(m), par_dims, order='F')]
ValueError: dims must have at least one value
´´´

**Comment by ahartikainen** [comment]
Ok, yes that is an error in plotting code, it assumes more than one chain, so it is a bug.

**Comment by hanyas** [comment]
Not sure if last commits are still WIP, if yes then please ignore this post. 

I pulled them anyway and I get the following error on the example I provided:

```
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "stanfit4anon_model_9c4d808afb937f9440712057e5cdb74a_2616328286471641134.pyx", line 511, in stanfit4anon_model_9c4d808afb937f9440712057e5cdb74a_2616328286471641134.StanFit4Model.plot
AttributeError: type object 'type' has no attribute 'lower'
```

**Comment by ahartikainen** [comment]
Yes they were WP.

Actually I forgot to test for "chain=1" situation, will do that later, but it should now work.

Also added keyword `cmap` so `cmap='viridis'` etc works. It can also use any `plt.cm.cmap_func` so colorcet cmaps can be used is needed.

**Comment by hanyas** [comment]
Yep both issues resolved. Single chain works. Thank you! I am really looking forward to the stable release!

**Comment by ahartikainen** [comment]
Ok, I'm thinking of merging this as is. (I will go over the code and send example figs here before the merge.)

Due to recent PRs in mcmcplotlib I think we should go forward with it as a backend. 

So let's merge this (2.18). Then wrap mcmcplotlib and add that first into the pystan.experimental and then if the everything works let's replace the current code with mcmcplotlib.

@ariddell any thoughts

**Comment by ariddell** [comment]
Looks great. Do we want to add a few superficial tests? Perhaps just call the various plot functions. This would at least allow us to verify that no exceptions are being raised.

**Comment by ahartikainen** [comment]
I'm not totally sure what to do with this.

I'm inclined to wait for the first release of the arviz (renamed mcmcplotlib) and wrap it as a plotting backend.

Any pointers? @ariddell

@aloctavodia,  does arviz have any release schedule?

**Comment by ariddell** [comment]
I'm +1 on using arviz/mcmcplotlib. Waiting sounds like a good idea to me.

**Comment by aloctavodia** [comment]
Not yet, probably the first release could be for the last days of Abril or first days of May and a second release during July/August. Is this OK with you? Do you want/need something faster?

**Comment by ahartikainen** [comment]
Closing this for now. I will make a new PR for wrapping ArviZ.

**Pull request closed without merging**